### PR TITLE
Add control code scf.forall to scf.for pass

### PR DESCRIPTION
**Pull request by jtuyls**
Adds a pass that converts control code scf.forall ops into scf.for ops. This enables more loop subsumption opportunities as scf.forall needs to be either fully subsumed into a DMA or not at all, while converting to scf.for allows the loop to be partially subsumed.

With this pass added, the 4096x512x512 matmul on 2x2 lowers to 6532 words of control code instead of 52228 before this pass. For this configuration with ukernel, the latency is now 10ms compared with 19ms earlier.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeForallToFor.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Converts `scf.forall` operations found within the provided op into nested
+/// `scf.for` operations.
+LogicalResult forallToFor(RewriterBase &rewriter, Operation *op) {
+  WalkResult res = op->walk([&](scf::ForallOp forallOp) {
+    rewriter.setInsertionPoint(forallOp);
+    if (succeeded(forallOp.promoteIfSingleIteration(rewriter))) {
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeForallToFor.cpp:31]: nit: add a test for this case

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeForallToFor.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      return WalkResult::advance();
+    }
+    if (failed(scf::forallToForLoop(rewriter, forallOp))) {
+      forallOp.emitOpError() << "failed to transform `scf.forall` to `scf.for`";
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeForallToFor.cpp:35]:
```suggestion
      forallOp.emitOpError() << "was not transformed from `scf.forall` to `scf.for`";
```

**Review by newling — COMMENTED** [review]
Looks good, just nit comments which can be ignored

Question though. Might the order of the scf.for loops matter? For example, if the final test only had `affine.apply  #map(%l)` might be it be better to have that for loop be the outer-most loop? Not sure if there's an easy way to do this analysis and subsequent loop inversion (and not sure if 'inversion' is the correct name)

**Review by newling — APPROVED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeForallToFor.cpp (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeForallToFor.cpp:31]: Jup, done

**Comment by jtuyls** [comment]
> Question though. Might the order of the scf.for loops matter? For example, if the final test only had `affine.apply #map(%l)` might be it be better to have that for loop be the outer-most loop? Not sure if there's an easy way to do this analysis and subsequent loop inversion (and not sure if 'inversion' is the correct name)

Yes, it does certainly matter for different shapes, but I think this should be done at the tiling level instead.

**Comment by yzhang93** [comment]
> > Question though. Might the order of the scf.for loops matter? For example, if the final test only had `affine.apply #map(%l)` might be it be better to have that for loop be the outer-most loop? Not sure if there's an easy way to do this analysis and subsequent loop inversion (and not sure if 'inversion' is the correct name)
> 
> Yes, it does certainly matter for different shapes, but I think this should be done at the tiling level instead.

Yes, we can control the order of loops when generating tiling strategy. The question is how to define such strategy. Previously we were thinking the simple way is to compare the M and N sizes, and make the outer loop with the smaller input size. But from the latest results, it looks like 4096x512x512 has better performance than 512x4096x512.

**Pull request merged**